pkg/reactor/document_controller.go

```diff
@@ -29,28 +30,25 @@ type docController struct {
 // NewDocumentController creates a controller for processing documents.
 func NewDocumentController(worker *DocumentWorker, workersCount int, failfast bool) DocumentController {
 	job := &jobs.Job{
-		ID:                        "Document",
-		MinWorkers:                workersCount,
-		MaxWorkers:                workersCount,
-		FailFast:                  failfast,
-		Worker:                    worker,
-		Queue:                     jobs.NewWorkQueue(2 * workersCount),
-		IsWorkerExitsOnEmptyQueue: true,
+		ID:         "Document",
+		MinWorkers: workersCount,
+		MaxWorkers: workersCount,
+		FailFast:   failfast,
+		Worker:     worker,
+		Queue:      jobs.NewWorkQueue(2 * workersCount),
 	}
+	job.SetIsWorkerExitsOnEmptyQueue(true)
 	return &docController{
 		jobs.NewController(job),
 		job,
 	}
 }
 func (d *docController) Shutdown() {
-	d.Controller.Shutdown()
 	// propagate the shutdown to the related download controller
-	d.Worker.(*DocumentWorker).NodeContentProcessor.GetDownloadController().Shutdown()
+	defer d.Worker.(*DocumentWorker).NodeContentProcessor.GetDownloadController().Shutdown()
+	klog.Warning("Shutting down Doc controller")
+	d.Controller.Shutdown()
```

Comment:
However, I'm missing the shutdown of the potentially configured `GitInfoController`.
Overall, controlling the shutdown from here smells. It should be orchestrated externally to keep the controllers independent. 
I suggest as a quick fix to add for now the optional controller shutdown in a different PR, and consider refactoring in the longer run.